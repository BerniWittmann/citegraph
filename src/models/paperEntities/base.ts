import { AuthorFields } from '@/models/paperEntities/author'
import { ConferenceFields } from '@/models/paperEntities/conference'
import { CountryFields } from '@/models/paperEntities/country'
import { InstitutionFields } from '@/models/paperEntities/institution'
import { PublicationFields } from '@/models/paperEntities/publication'
import { PublisherFields } from '@/models/paperEntities/publisher'
import { RecordFields } from '@/models/paperEntities/record'
import { constructQuery, PaperEntityQueryParameters } from '@/models/paperEntities/query'

export interface PaperEntityBaseFields {
  id: string;
}

export default abstract class PaperEntity implements PaperEntityBaseFields {
  id: string;
  static readonly key: string;
  static readonly schemaName: string;
  static readonly queryName: string;
  static readonly displayedColumns: Array<PaperEntityTableColumn>;
  static readonly queryFields: string;

  protected constructor ({ id }: PaperEntityBaseFields) {
    this.id = id
  }

  static getQuery (params: PaperEntityQueryParameters): string {
    return constructQuery(this.queryName, this.schemaName, params, this.queryFields)
  }
}

export interface PaperEntityTableColumn {
  text: string;
  value: string;
  align?: 'start' | 'center' | 'end'
  sortable?: boolean
  filterable?: boolean
  divider?: boolean
  class?: string | string[]
  width?: string | number
  filter?: (value: any, search: string, item: any) => boolean
  sort?: (a: any, b: any) => number
}

export type PaperEntityFields = PaperEntityBaseFields & (
  AuthorFields &
  ConferenceFields &
  CountryFields &
  InstitutionFields &
  PublicationFields &
  PublisherFields &
  RecordFields
)
