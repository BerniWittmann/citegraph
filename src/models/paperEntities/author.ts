import PaperEntity, { PaperEntityBaseFields, PaperEntityTableColumn } from './base'
import AuthorTransformer from '@/transformers/AuthorTransformer'

export interface AuthorFields {
  firstName: string
  lastName: string
  countRecords: number
}

export default class Author extends PaperEntity implements AuthorFields {
  static key: string = 'author'
  static schemaName: string = 'author'
  static queryName: string = 'Authors'
  static queryFields: string = `id,
          first_name,
          last_name,
          number_citations`
  static mutationFields: Array<string> = ['first_name', 'last_name']
  static displayedColumns: Array<PaperEntityTableColumn> = [{
    text: 'project.explore.table.headers.author.first_name',
    value: 'firstName',
    sortable: true,
    filterable: true,
    displayedByDefault: true
  }, {
    text: 'project.explore.table.headers.author.last_name',
    value: 'lastName',
    sortable: true,
    filterable: true,
    displayedByDefault: true
  }, {
    text: 'project.explore.table.headers.author.count_records',
    value: 'countRecords',
    sortable: true,
    align: 'center'
  }]
  static transformer = AuthorTransformer
  firstName: string
  lastName: string
  countRecords: number

  constructor ({ id, firstName, lastName, countRecords }: PaperEntityBaseFields & AuthorFields) {
    super({ id })
    this.firstName = firstName
    this.lastName = lastName
    this.countRecords = countRecords
  }
}
