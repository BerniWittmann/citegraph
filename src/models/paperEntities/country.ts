import PaperEntity, { PaperEntityBaseFields, PaperEntityTableColumn } from './base'

export interface CountryFields {
  name: string
  countRecords: number
}

export default class Country extends PaperEntity implements CountryFields {
  static key: string = 'country'
  static schemaName: string = 'country'
  static queryName: string = 'Countries'
  static queryFields: string = `name,
          numberCitations`
  static displayedColumns: Array<PaperEntityTableColumn> = []
  name: string
  countRecords: number

  constructor ({ id, name, countRecords }: PaperEntityBaseFields & CountryFields) {
    super({ id })
    this.name = name
    this.countRecords = countRecords
  }
}
