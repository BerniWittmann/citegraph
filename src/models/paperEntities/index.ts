import Author from './author'
import Conference from './conference'
import Country from './country'
import Institution from './institution'
import Publication from './publication'
import Publisher from './publisher'
import Record from './record'
import PaperEntity, { PaperEntityTableColumn, PaperEntityFields } from './base'

export const entities = [
  Record,
  Author,
  Institution,
  Country,
  Publication,
  Publisher,
  Conference
]

export const entityKeysMap = {
  [Record.key]: Record,
  [Author.key]: Author,
  [Institution.key]: Institution,
  [Country.key]: Country,
  [Publication.key]: Publication,
  [Publisher.key]: Publisher,
  [Conference.key]: Conference
}

export const entityKeys: Array<string> = entities.map((entity) => entity.key)

export type PaperEntityTableColumn = PaperEntityTableColumn

export type PaperEntity = PaperEntity

export type PaperEntityFields = PaperEntityFields

export default {
  Author,
  Conference,
  Country,
  Institution,
  Publication,
  Publisher,
  Record
}
