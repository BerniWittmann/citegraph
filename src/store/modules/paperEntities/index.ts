import { Module } from 'vuex'
import { PaperEntitiesState } from './types'
import { RootState } from '@/store/types'
import { state } from './state'
import { actions } from './actions'
import { getters } from './getters'
import { mutations } from './mutations'

const namespaced: boolean = true

export const paperEntities: Module<PaperEntitiesState, RootState> = {
  namespaced,
  state,
  getters,
  actions,
  mutations
}
