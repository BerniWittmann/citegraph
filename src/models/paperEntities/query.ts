import { PaperEntityFields } from '@/models/paperEntities/base'

export interface PaperEntityQueryParameters {
  projectId: number
  entityType: string
  perPage?: number
  pageOffset?: number
  filter?: string
  sortBy?: string,
  id?: string
}

export type PaperEntityQueryResponse = PaperEntityFields

function createQueryParams (params: PaperEntityQueryParameters): string {
  if (!params.id && !params.perPage && !params.pageOffset && !params.filter && !params.sortBy) return ''
  if (params.id) {
    return `(id: ${params.id})`
  }
  const args = [
    params.filter ? `filter: ${params.filter}` : '',
    params.sortBy ? `orderBy: ${params.sortBy}` : ''
  ]
  if (params.perPage !== undefined && params.pageOffset !== undefined) {
    args.push(`first: ${params.perPage}`)
    if (params.pageOffset > 0) {
      args.push(`skip: ${params.pageOffset * params.perPage}`)
    }
  }
  return '(' + args.filter((arg) => arg.length > 0).join(' ') + ')'
}

export function constructQuery (queryName: string, schemaName: string, params: PaperEntityQueryParameters, fields: string): string {
  return `{
      ${queryName}${createQueryParams(params)} {
        count
        ${schemaName} {
          ${fields}
        }
      }
    }`
}
