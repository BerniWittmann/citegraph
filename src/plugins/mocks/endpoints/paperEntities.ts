import db from '../db'
import MockAdapter from 'axios-mock-adapter'
import { RecordFields } from '@/models/paperEntities/record'
import { setupMocks as setupGraphQLMocks } from './graphql-helpers/endpoint'
import { QueryConfiguration } from './graphql-helpers/queries'
import { AuthorFields } from '@/models/paperEntities/author'
import { CountryFields } from '@/models/paperEntities/country'

const records = db.getCollection('records')
const authors = db.getCollection('authors')
const countries = db.getCollection('countries')

const queryResponseConfiguration: Record<string, QueryConfiguration> = {
  'Records': {
    schemaName: 'record',
    queryName: 'Records',
    collection: records,
    filterFunction: recordsFilter
  },
  'Authors': {
    schemaName: 'author',
    queryName: 'Authors',
    collection: authors,
    filterFunction: authorsFilter
  },
  'Countries': {
    schemaName: 'country',
    queryName: 'Countries',
    collection: countries,
    filterFunction: countriesFilter
  }
}

function recordsFilter (obj: RecordFields, filter: string) {
  return obj.title.toLowerCase().includes(filter) ||
    obj.year.toString().includes(filter) ||
    obj.keywords.map(keyword => keyword.toLowerCase()).join('').includes(filter) ||
    obj.authors.map(author => (author.firstName + ' ' + author.lastName).toLowerCase()).join(';').includes(filter)
}

function authorsFilter (obj: AuthorFields, filter: string) {
  return obj.firstName.toLowerCase().includes(filter) ||
    obj.lastName.toLowerCase().includes(filter)
}

function countriesFilter (obj: CountryFields, filter: string) {
  return obj.name.toLowerCase().includes(filter)
}

export function setupMocks (mock: MockAdapter) {
  setupGraphQLMocks(mock, queryResponseConfiguration)
}
