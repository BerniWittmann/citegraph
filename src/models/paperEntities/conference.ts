import PaperEntity, { PaperEntityBaseFields, PaperEntityTableColumn } from './base'
import { createFilterAndPaginationForQuery } from '@/models/paperEntities/query'

export interface ConferenceFields {
  name: string
  countRecords: number
}

export default class Conference extends PaperEntity implements ConferenceFields {
  static key: string = 'conference'
  static schemaName: string = 'conference'
  static queryName: string = 'Conferences'
  static displayedColumns: Array<PaperEntityTableColumn> = []
  name: string
  countRecords: number

  constructor ({ id, name, countRecords }: PaperEntityBaseFields & ConferenceFields) {
    super({ id })
    this.name = name
    this.countRecords = countRecords
  }

  static getQuery (perPage?: number, pageOffset?: number, filter?: string): string {
    return `{
      ${Conference.queryName}${createFilterAndPaginationForQuery(perPage, pageOffset, filter)} {
        count
        ${Conference.schemaName} {
          name,
          numberCitations
        }
      }
    }`
  }
}
