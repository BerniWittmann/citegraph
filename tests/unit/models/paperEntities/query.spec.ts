import { constructQuery } from '@/models/paperEntities/query'

describe('models/paperEntities/query.ts', () => {
  describe('it has a function to create a query', () => {
    const baseParams = {
      projectId: 42,
      entityType: 'record'
    }
    it('creates a query', () => {
      expect(constructQuery('Record', 'record', { ...baseParams, perPage: 7, pageOffset: 2, filter: 'test', sortBy: 'title_ASC' }, 'id, title, name')).toMatchSnapshot()
      expect(constructQuery('Record', 'record', { ...baseParams, id: '12' }, 'id, title, count')).toMatchSnapshot()
    })

    describe('it can query for a single entity by id', () => {
      it('adds an id to the query', () => {
        expect(constructQuery('Record', 'record', { ...baseParams, id: '12' }, 'id')).toContain('(id: 12)')
      })
    })

    describe('it can query for multiple with pagination and filter', () => {
      it('returns an empty string if no parameters are given', () => {
        expect(constructQuery('Record', 'record', { ...baseParams }, 'id')).not.toContain('(')
      })

      it('returns only the filter if only the filter is given', () => {
        expect(constructQuery('Record', 'record', { ...baseParams, filter: 'test' }, 'id')).toContain('(filter: test)')
      })

      it('returns only the sorting parameter if only the sorting parameter is given', () => {
        expect(constructQuery('Record', 'record', { ...baseParams, sortBy: 'title_DESC' }, 'id')).toContain('(orderBy: title_DESC)')
      })

      it('returns only the pagination if only the pagination is given', () => {
        expect(constructQuery('Record', 'record', { ...baseParams, perPage: 10, pageOffset: 1 }, 'id')).toContain('(first: 10 skip: 10)')
        expect(constructQuery('Record', 'record', { ...baseParams, perPage: 7, pageOffset: 2 }, 'id')).toContain('(first: 7 skip: 14)')
        expect(constructQuery('Record', 'record', { ...baseParams, perPage: 7, pageOffset: 0 }, 'id')).toContain('(first: 7)')
      })

      it('if all parameters are given it returns all parameters ', () => {
        expect(constructQuery('Record', 'record', { ...baseParams, perPage: 7, pageOffset: 2, filter: 'test', sortBy: 'title_ASC' }, 'id')).toContain('(filter: test orderBy: title_ASC first: 7 skip: 14)')
      })
    })
  })
})
