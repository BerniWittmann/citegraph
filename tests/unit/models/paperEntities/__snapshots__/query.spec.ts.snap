// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models/paperEntities/query.ts it has a function to create a query creates a query 1`] = `
"{
      Record(filter: test orderBy: title_ASC first: 7 skip: 14) {
        count
        record {
          id, title, name
        }
      }
    }"
`;

exports[`models/paperEntities/query.ts it has a function to create a query creates a query 2`] = `
"{
      Record(id: 12) {
        count
        record {
          id, title, count
        }
      }
    }"
`;
